### PR TITLE
Fix segfault when passing empty StringIO/File object to Font

### PR DESCRIPTION
**Pull request by MyreMylar**
Related to #2295

Also fixes freetype submodule's `Font` to track submodule generation on `__new__` instead of on `__init__` for more reliable performance of the unit tests with uninitialized Font objects. Previously I was getting different results running on the CI versus locally.

### CODE REVIEW

**Comment by MyreMylar** [comment]
I also fixed two unrelated freetype module tests here that I noticed were failing locally while testing - but were not failing on the CI. 

At some point a fix that was added for a quit() and init() segfault and the way the tests multithread on the CI causes these two particular tests which I changed, which use a class level `__new__` constructor, to give one exception on the CI (that the module has been quit before the 'font' is created) and a different exception locally (the exceptions these tests were originally designed to capture). 

It is not a real functional change but it means we can get these tests to pass both locally and on the CI.

File: test/freetype_test.py [hunk 1]
```diff
@@ -544,7 +549,16 @@ def test_freetype_Font_name(self):
         f = self._TEST_FONTS["fixed"]
         self.assertEqual(f.name, "Inconsolata")
 
-        self.assertRaises(RuntimeError, lambda: nullfont().name)
+        with self.assertRaises(Exception) as cm:
+            # this test raises an attribute error locally but
+            # a pygame.error on the CI - I assume this is multi-threading
```
Comment (dr0id) [inline test/freetype_test.py:554]: Are sure about this assumption?

I think I changed that in PR #2371 not sure why it behaves this way.

Comment (MyreMylar) [reply test/freetype_test.py]: I had another idea, about adding the freetype generation to all Font Objects - whether they succeed in loading or not - to see if that will erase the difference between CI and local for these tests.

Comment (MyreMylar) [reply test/freetype_test.py]: OK, with a bit of debugging it looks like the answer is that we should be setting the freetype generation for a font on `__new__` rather than on `__init__` so it lines up properly for uninitialized fonts. It worked OK when everything lined up to be zero anyway (no inits or quits of the module) but whenever that was not true this was breaking because the module generation variable was only set to anything other than 0 for initialised font objects.

I added a bunch of stuff to the error messages to debug which I'm going to unwind to make the diff smaller.

File: test/freetype_test.py [hunk 2]
```diff
@@ -1349,7 +1356,10 @@ def test_freetype_Font_resolution(self):
 
     def test_freetype_Font_path(self):
         self.assertEqual(self._TEST_FONTS["sans"].path, self._sans_path)
-        self.assertRaises(pygame.error, lambda: nullfont().path)
+
+        with self.assertRaises(AttributeError):
+            nullfont = ft.Font.__new__(ft.Font)
+            nullfont.path
```
Comment (ankith26) [inline test/freetype_test.py:1362]: Just to confirm, the init generation changes + these test changes are not related to the segfault, right?

Comment (MyreMylar) [reply test/freetype_test.py:1362]: Nope, it may be that my adjustments to freetype Font init exposed the issue on the CI, but it the underlying issue is unrelated. 

Basically the tests in question here are using an 'uninitialized font' which means one where you don't call __init__() and the generation variable was only being set in __init__() so unsurprisingly if anything happened with quit() at the module level the uninitialized font's generation value (which was previously always 0) would be wrong and the tests would start spitting out the module has been quit error instead of relevant errors.

**Review by ankith26 — APPROVED** [review]
Thanks! Approving this PR because this LGTM, but I wonder if this is something that should also be fixed upstream on `SDL_ttf` if this is an issue on their end?

**Review by ankith26 — COMMENTED** [review]
Another improvement idea, we should use `SDL_RWsize(source)` instead of doing `source->size(source)`. Shorter and also looks more robust

**Review by dr0id — APPROVED** [review]

**Comment by MyreMylar** [comment]
> Another improvement idea, we should use `SDL_RWsize(source)` instead of doing `source->size(source)`. Shorter and also looks more robust


> This function is available since SDL 2.0.10.
https://wiki.libsdl.org/SDL2/SDL_RWsize
